File: collectData.js
const fs = require('fs');
const req = require('request');
const unzipper = require('unzip-stream');
const ffmpeg = require('fluent-ffmpeg');

function getBeatMapSet(curCursor, filter) {
    var info;
    return new Promise((resolve, reject) => {
        // console.log("https://osu.ppy.sh/beatmapsets/search" + filter + curCursor);
        req("https://osu.ppy.sh/beatmapsets/search" + filter + curCursor, (error, response, body) => {
            info = body;
            resolve(info);
        });
    });
};

function dlBeatMap(beatmapset) {
    let id = beatmapset.id;
    return new Promise((resolve, reject) => {
        let path = 'data/' + id + "/";
        // Clean folder first
        fs.readdir(path, (err, files) => {
            if (!err) {
                for (const file of files) {
                    fs.unlink(path + file, err => {
                      if (err) throw err;
                    });
                  }
            }
            
            // Download the beatmap from osu
            req("https://osu.ppy.sh/d/" + id, () => {
                unzipClean(beatmapset);
            }).pipe(fs.createWriteStream(path + id + ".zip"));
        });
        resolve();
    });
};

function unzipClean(beatmapset) {
    let id = beatmapset.id;
    let path = 'data/' + id + "/";
    fs.createReadStream(path + id+ ".zip")
    // Unzip the file
    .pipe(unzipper.Parse())
    .on('entry', function (entry) {
        const fileName = entry.path;
        // Only keep mp3 and osu files
        if (fileName.endsWith(".mp3")) {
            entry.pipe(fs.createWriteStream('data/' + id + "/" + id + ".mp3"));
        }
        else if (fileName.endsWith(".osu")) {
            entry.pipe(fs.createWriteStream('data/' + id + "/" + id + fileName));
        }
        else {
            entry.autodrain();
        }
    })
    .on('close', () => {
        fs.readdir(path, (err, files) => {
            files.forEach((file) => {
                // Delete zip files
                if (file.endsWith(".zip")) {
                    fs.unlink(path + file, (err) => {
                        if (err) throw err;
                    });
                }
                // Convert mp3 files to wav files
                else if (file.endsWith(".mp3")) {
                    mp3towav(path, file);
                }
                // Only keep osumania osu files, e.g. Mode: 3
                else if (file.endsWith(".osu")) {
                    filterOsuManiaFiles(path, file, id);
                }
            })
        })
    });
}

function filterOsuManiaFiles(path, file, id) {
    console.log(path + file + " : " + id);
    fs.readFile(path + file, (err, data) => {
        if (data.indexOf("Mode: 3") < 0) {
            fs.unlink(path + file, (err) => {
                if (err) throw err;
            });
        }
        else {
            // Grab the beatmapId and place the difficulty as part of the file name
            let dataStr = data.toString("utf-8");
            let it = dataStr.indexOf("BeatmapID:") + 10;
            let beatmapId = "";
            while (!isNaN(parseInt(dataStr[it]))) {
                beatmapId += dataStr[it];
                ++it;
            }
            console.log(beatmapId);
            var beatmap = beatmapset.beatmaps.filter((beatmap) => {
                return beatmap.id == beatmapId;
            });

            if (beatmap.length != 1) throw "Found " + beatmap.length + " beatmaps for " + beatmapId;
            beatmap = beatmap[0];

            fs.unlink(path + id + "[" + beatmap.difficulty_rating + "].osu", (err) => {
                fs.renameSync(path + file, path + id + "[" + beatmap.difficulty_rating + "].osu", (err) => {});
            });
        }
    })
}

function mp3towav(path, file) {
    return new Promise((resolve, reject) => {
        let wavFileName = file.replace(".mp3", ".wav");
        ffmpeg(path + file)
        .toFormat('wav')
        .on('error', (err) => {
            console.log('Error occurred on converting mp3 to wav: ' + err.message);
        })
        .on('end', () => {
            fs.unlink(path + file, (err) => {});
        })
        .save(path + wavFileName)
        resolve();
    })
}

async function main() {
    let numData = 10;

    if (process.argv.length > 2) {
        numData = parseInt(process.argv[2]);
        if (isNaN(numData)) throw "Parameter is not a number!";
    }

    console.log("Finding " + numData + " Mania Songs");

    const filter = "?m=3&sort=plays_desc&s=any";
    let curCursor = "";

    var data;
    let max = -1;
    let numFound = 0;
    let numSongs = 0;
    do {
        data = JSON.parse(await getBeatMapSet(curCursor, filter));
        console.log(data.cursor);
        for (beatmapset of data.beatmapsets) {
            if (beatmapset.availability.download_disabled) continue;
            let hasMania = false;
            for (beatmap of beatmapset.beatmaps) {
                if (beatmap.mode_int == 3) {
                    hasMania = true;
                    break;
                }
            }
            if (hasMania) {
                console.log(beatmapset.title + ": " + beatmapset.id + " : " + beatmapset.play_count);
                fs.mkdir('data/' + beatmapset.id, {recursive:true}, (err) =>{});
                await dlBeatMap(beatmapset);
                if (++numFound >= numData) break;
            }
        }
        numSongs += data.beatmapsets.length;
        curCursor = "&cursor%5Bplay_count%5D=" + data.cursor.play_count + "&cursor%5B_id%5D=" + data.cursor._id;

        if (max < 0) {
            max = data.total;
            console.log(max);
        }
    } while (numFound < numData && max > (numSongs));

    console.log("Last Cursor: " + JSON.stringify(data.cursor));
}

main();

File: collectData.py
import urllib.request
import requests
import re
import json
import os
import sys
import zipfile
import shutil
import time

def getBeatMapSet(curCursor, urlFilter):
    response = requests.get("https://osu.ppy.sh/beatmapsets/search" + urlFilter + curCursor)
    return response.json()

def dlBeatMap(beatmapset, osuSessIds):
    bmsId = beatmapset['id']
    bmsPath = "data/{}/".format(bmsId)
    # Clean folder first
    for fileName in os.listdir(bmsPath):
        filePath = bmsPath + fileName
        try:
            if os.path.isfile(filePath) or os.path.islink(filePath):
                os.unlink(filePath)
            elif os.path.isdir(filePath):
                shutil.rmtree(filePath)
        except Exception as e:
            print("Error when deleting {}. Reason: {}".format(filePath, e))

    # Download the beatmap from osu

    # response = requests.get(
    #     "https://old.ppy.sh/d/{}".format(bmsId),
    #     headers={'osu_session': osuSessIds},
    #     stream=True
    # )
    osuSessIds = "eyJpdiI6ImdEZGFoMmtZOEV1TzhSRVwvUW9wME1nPT0iLCJ2YWx1ZSI6IkZ6dEErVnRncXRYYjNveG9NWDZJTVgyeU9QQmZaaDhnSHhvNFhUMWpLYjBCaWkyU2lcL2o1UklFQnBNOHFpcFRyMVBNcVBVSmFjNUZKQXNZdEpFTThjUT09IiwibWFjIjoiN2EwYWU3YTgzMWM4OTIwZWRiMDAzOWMwYmEyMjRkNzkyNWIzNzVkNjc4NTk0MTdjZGM5YThjMmE0NjNjY2FlYiJ9"
    quota_chk = requests.get(
        "https://osu.ppy.sh/home/download-quota-check".format(bmsId),
        headers={'cookie': 'osu_session={}'.format(osuSessIds)},
        stream=True
    )
    print("STATUS CODE {}: {}".format(bmsId, quota_chk.status_code))
    response = requests.get(
        "https://osu.ppy.sh/beatmapsets/{}/download".format(bmsId),
        headers={'cookie': 'osu_session={}'.format(osuSessIds)},
        stream=True
    )
    print("STATUS CODE {}: {}".format(bmsId, response.status_code))
    with open(bmsPath + "{}.zip".format(bmsId), 'wb') as beatmap:
        for chunk in response.iter_content(chunk_size=512 * 1024):
            if chunk:
                beatmap.write(chunk)


if __name__ == "__main__":
    numData = 10

    if (len(sys.argv) > 1):
        numData = int(sys.argv[1])

    print("Finding {} Mania Songs".format(numData))

    # Authentication
    
    cred = open("cred.txt", "r")
    credLines = cred.readlines()
    osuUser = credLines[0]#input('Enter your osu username: ')
    osuPass = credLines[1]#input('Enter your osu password: ')

    authentication_url = 'https://osu.ppy.sh/session'
    payload = {
        'action': 'login',
        'username': osuUser,
        'password': osuPass,
        'redirect': 'index.php',
        'sid': '',
        'login': 'Login'
    }

    response = requests.post(authentication_url, data=payload)
    print(response.headers['Set-Cookie'])#.split("; "))
    osuSessIds = list(filter(lambda x: "osu_session=" in x ,response.headers['Set-Cookie'].split("; ")))[1]
    osuSessIds = osuSessIds.split("osu_session=")[1]
    print("STATUS_CODE")
    assert response.status_code == 200, "Please check your credentials, (in cred.txt)"

    urlFilter = "?m=3&sort=plays_desc&s=any"
    curCursor = ""

    maxSongs = sys.maxsize
    numFound = 0
    numSongs = 0
    while (numFound < numData and maxSongs > (numSongs)):
        data = getBeatMapSet(curCursor, urlFilter)

        # Search through all beatmapsets
        for beatmapset in data['beatmapsets']:
            if (beatmapset['availability']['download_disabled']):
                continue
            hasMania = False
            for beatmap in beatmapset['beatmaps']:
                if (beatmap['mode_int'] == 3):
                    hasMania = True
                    break
            if (hasMania):
                print("{}: {} : {}".format(beatmapset['title'], beatmapset['id'], beatmapset['play_count']))
                beatmapSetPath = 'data/{}'.format(beatmapset['id'])
                if not os.path.exists(beatmapSetPath):
                    os.mkdir(beatmapSetPath)

                dlBeatMap(beatmapset, osuSessIds)
                numFound += 1
                if (numFound >= numData):
                    break
            
            if (maxSongs == sys.maxsize):
                maxSongs = data['total']
                print("Max Songs: {}".format(maxSongs))
        
        numSongs += len(data['beatmapsets'])
        curCursor = "&cursor%5Bplay_count%5D={}&cursor%5B_id%5D={}".format(data['cursor']['play_count'], data['cursor']['_id']) 
        
        print(data['cursor'])